src/dune_pkg/rev_store.ml

```diff
@@ -391,22 +391,88 @@ module Entry = struct
   ;;
 end
 
-module Submodule = struct
-  (* a submodule in [.gitmodules] can also have a [branch] but given we only
-     need to resolve the commit object, we don't have to care about the
-     tracking branch *)
+module At_rev = struct
+  type repo = t
+
   type t =
-    { path : Path.Local.t
+    { repo : repo
+    ; revision : Rev.t
     ; source : string
+    ; files : File.Set.t
     }
 
-  let parse lines =
-    match Git_config_parser.parse lines with
-    | Error err ->
-      (* CR-rgrinberg: the loc needs to be pulled from the git URL *)
-      User_error.raise [ Pp.textf "Failed to parse submodules: %s" err ]
-    | Ok cfg ->
-      List.filter_map cfg ~f:(fun { name; arg = _; bindings } ->
+  module Config = struct
+    type bindings = string * string
+
+    type section =
+      { name : string
+      ; arg : string option
+      ; bindings : bindings list
+      }
+
+    type t = section list
+
+    module KV = struct
+      module T = struct
+        type t = string * string option
+
+        let compare = Tuple.T2.compare String.compare (Option.compare String.compare)
+        let to_dyn = Tuple.T2.to_dyn Dyn.string (Dyn.option Dyn.string)
+      end
+
+      include Comparable.Make (T)
+    end
+
+    let config { dir } revision path : t Fiber.t =
+      let (Rev.Rev rev) = revision in
+      let location = [ "--blob"; sprintf "%s:%s" rev (Path.Local.to_string path) ] in
+      let command = [ "config"; "--list" ] @ location in
+      let failure_mode =
+        Process.Failure_mode.Accept
+          (Predicate.create (fun x -> Int.equal x 0 || Int.equal x 128))
+      in
+      let git = Lazy.force Vcs.git in
+      let+ content =
+        let stderr_to = make_stderr () in
+        Process.run_capture_lines ~dir ~display:Quiet ~stderr_to failure_mode git command
+      in
+      match content with
+      | Error _ -> []
+      | Ok content ->
+        content
+        |> List.fold_left ~init:KV.Map.empty ~f:(fun acc line ->
+          match String.lsplit2 ~on:'=' line with
+          | None -> acc
+          | Some (key, value) ->
+            (match String.lsplit2 ~on:'.' key with
+             | None -> acc
+             | Some (section, key) ->
+               let arg, binding =
+                 match String.rsplit2 ~on:'.' key with
```

Comment:
I suppose that the `rsplit2` takes care of parsing things like `branch.compat-5.0-dune-2.9.remote=origin`, but tests for these kinds of things would be useful

Reply:
Indeed. I've added some tests for this. It's sort of ugly that this needs to be exposed to be tested with expect tests, would be nice to avoid that if there was a better way (I could stuff it into a `Private` module but that's also just a workaround).

Reply:
One alternative is to make a test that has a fake `git` binary in scope. But that's not necessarily better.

Reply:
True; haven't thought of that. I think the code is now fine-ish. If the processing was more complicated to the point that returning a tuple wouldn't be feasible a test `git` binary would indeed be the better alternative.